ui/raidboss/emulator/data/network_log_converter/LineEvent0x00.ts

```diff
@@ -1,14 +1,24 @@
 import LineEvent from './LineEvent';
+import LogRepository from './LogRepository';
+
+export type LineEvent0x00Parts = [
+  string, // 0: ID
+  string, // 1: Timestamp
+  string, // 2: Type
+  string, // 3: Speaker
+  string, // 4: Message
+  string, // 5: Checksum
+];
 
 // Chat event
 export class LineEvent0x00 extends LineEvent {
-  constructor(repo, line, parts) {
+  constructor(repo: LogRepository, line: string, public parts: LineEvent0x00Parts) {
```

Comment:
`LineEvent0x00Parts` seems a bit suspicious.  I think `parts` comes from `split('|')` and from data off disk (possibly constructed, corrupted, incorrect).  So, it really seems like we don't know anything more about it other than `string[]`.  I don't think it makes sense to assert via `as` that the incoming parameter has 6 string fields, as that assertion doesn't add any type safety.   It seems safer to say `parts: string[]` and just handle the potentially undefined parts as needed.  If it

Additionally, I think it should be `parts: string[]` and not `public parts: string[]`, as `LineEvent` already stores parts and `LineEvent0x00` doesn't need to also?

Reply:
This is a typecasting adjustment. Without the custom type, the array is actually read as `(string | undefined)[]` for some reason, so it results in this:

```typescript
  constructor(repo: LogRepository, line: string, parts: string[]) {
    super(repo, line, parts);

    this.id = parts[2].toUpperCase();
    this.name = parts[3];

    this.abilityId = parts[4].toUpperCase();
    this.abilityName = parts[5];
  }
```

```
[tsl] ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x17.ts(26,15)
      TS2532: Object is possibly 'undefined'.

ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x17.ts
27:4-13
[tsl] ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x17.ts(27,5)
      TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x17.ts
29:21-29
[tsl] ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x17.ts(29,22)
      TS2532: Object is possibly 'undefined'.

ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x17.ts
30:4-20
[tsl] ERROR in C:\Users\valar\Source\cactbot2\ui\raidboss\emulator\data\network_log_converter\LineEvent0x17.ts(30,5)
      TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
```

The `public` assignment is to avoid this issue in the `convert` function, as well.

Reply:
I'm trying to argue that TypeScript is correct, and that it is potentially undefined.  Asserting that it is defined does not make it so.

Reply:
Throwing an `assert` to check that it's actually the correct length and all values are string doesn't fix tsc complaining about it potentially being undefined. I'm not sure what would fix this satisfactorily. Does the constructor become 200+ lines to handle the potential undefined values (in the case of 0x15 lines for example) to work around the issue?

Adding a check `if parts length is less than expected or anything in it isn't a string` fixes the logic issue but tsc still complains that it could be undefined even though we're checking that it's not already.

Reply:
Yeah, agreed.  A length check is not sufficient for TypeScript.

What about nullish coalescing?  I don't think that would add a ton of extra lines.

```typescript
export class LineEvent0x00 extends LineEvent {
  constructor(repo: LogRepository, line: string, parts: string[]) {
    super(repo, line, parts);
    if (this.parts.slice(4).join('|').split('\u001f\u001f').length > 1)
      this.invalid = true;
  }

  convert(_: LogRepository): void {
    const type = this.parts[2] ?? '';
    // If speaker is blank, it's excluded from the converted line
    const speaker = this.parts[3] && this.parts[3] !== '' ? `${this.parts[3]}:` : '';
    this.convertedLine =
      this.prefix() +
      type + ':' +
      speaker +
      this.parts.slice(4, this.parts.length - 1).join('|').trim();
    this.convertedLine = LineEvent00.replaceChatSymbols(this.convertedLine);
  }
```

Reply:
That seems to resolve the issues properly, but it's still going to be ugly for some of the classes. For example, these cases:
```typescript
    this.abilityId = parts[2].toUpperCase();
...
    this.stacks = parseInt(parts[9], 16);
...
    this.convertedLine = this.prefix() + this.targetId + ':' + this.targetName +
    ' gains the effect of ' + this.abilityName +
    ' from ' + this.fallbackResolvedTargetName +
    ' for ' + this.parts[4] +
    ' Seconds.' + stackCountText;
```

Compare to:

```typescript
    this.abilityId = (parts[2] ?? '').toUpperCase();
...
    this.stacks = parseInt(parts[9] ?? '0', 16);
...
    this.convertedLine = this.prefix() + this.targetId + ':' + this.targetName +
    ' gains the effect of ' + this.abilityName +
    ' from ' + this.fallbackResolvedTargetName +
    ' for ' + (this.parts[4] ?? '') +
    ' Seconds.' + stackCountText;
```

Reply:
Personally, I think nullish coalescing is better than using `as`.  There's probably cases right now where the raid emulator will throw an exception if you have some log with a corrupted line in it (such that the line is truncated and leaves some field as undefined), so I think this is a robustness improvement even if it makes the code slightly more awkward.

* `this.abilityId = parts[2]?.toUpperCase() ?? '';` is slightly more readable, as an alternative
* `this.stacks = parseInt(parts[9] ?? '0', 16);` seems fine (or alternatively make some `EmulatorCommon.parseHex(str?: string)` helper?)
* `(this.parts[4] ?? '')` can get a named temporary, like the previous comment, which I think is a readability improvement over the previous code

Reply:
I'm working on cleaning this up now, but I have a follow up question. In the case that a log line is invalid such that it can't realistically be used in the emulator, should we mark it as invalid so that it's ignored? That would be a change to current behavior so it might be better looked into as a future PR?

Reply:
Yeah, I think that'd be a nice follow-up.

Reply:
There is a trick in typescript to check if an array has a constant length, which is to add a `is-check` function, for expample:
```typescript
function isLineEvent0x00Parts(array: string[]): array is LineEvent0x00Parts {
  return array.length === 6;
}
```

I think this way would be better?